platforms/software/ivy/src/main/java/org/gradle/api/publish/ivy/tasks/PublishToIvyRepository.java

```diff
@@ -127,30 +119,28 @@ private static IvyPublicationInternal toPublicationInternal(IvyPublication publi
     /**
      * The repository to publish to.
      *
-     * @return The repository to publish to
+     * Only instances of DefaultIvyArtifactRepository are supported
      */
     @Internal
-    @ToBeReplacedByLazyProperty
+    @NotToBeReplacedByLazyProperty(because = "we need a better way to handle this, see https://github.com/gradle/gradle/pull/30665#pullrequestreview-2329667058")
     public IvyArtifactRepository getRepository() {
         return repository.get();
     }
 
-    @Nested
-    @Optional
-    Property<Credentials> getCredentials() {
-        return credentials;
-    }
-
     /**
      * Sets the repository to publish to.
      *
      * @param repository The repository to publish to
      */
     public void setRepository(IvyArtifactRepository repository) {
-        this.repository.set((DefaultIvyArtifactRepository) repository);
-        this.credentials.set(((DefaultIvyArtifactRepository) repository).getConfiguredCredentials());
+        this.repository.set(repository);
```

Comment:
PublishToMavenRepository uses the previous pattern, where the property is  of the default implementation type, and we cast in the setter. I prefer that since this method now fails lazily if you pass the wrong type.